python/ffsim/states/slater.py

```diff
@@ -0,0 +1,271 @@
+        bitstring_type: The desired type of bitstring output.
+        seed: A seed to initialize the pseudorandom number generator.
+            Should be a valid input to ``np.random.default_rng``.
+
+    Returns:
+        A 2D Numpy array with samples of electronic configurations.
+        Each row is a sample.
+    """
+    rng = np.random.default_rng(seed)
+
+    if isinstance(nelec, int):
+        # spinless case
+        norb, _ = rdm.shape
+
+        if orbs is None:
+            orbs = range(norb)
+
+        if nelec == 0:
+            strings = [0 for i in range(shots)]
+        elif nelec == norb:
+            strings = [sum(1 << j for j in range(norb)) for i in range(shots)]
+        else:
+            strings = _sample_spinless_direct(rdm, nelec, shots, rng)
+
+        strings = restrict_bitstrings(strings, orbs, bitstring_type=BitstringType.INT)
+
+        return convert_bitstring_type(
+            strings,
+            BitstringType.INT,
+            bitstring_type,
+            length=len(orbs),
+        )
+
+    else:
+        rdm_a, rdm_b = rdm
+
+        n_a, n_b = nelec
+        norb, _ = rdm_a.shape
+
+        if orbs is None:
+            orbs = (range(norb), range(norb))
+        orbs_a = orbs[0]
+        orbs_b = orbs[1]
+
+        if n_a == 0:
+            strings_a = [0 for i in range(shots)]
+        elif n_a == norb:
+            strings_a = [sum(1 << j for j in range(norb)) for i in range(shots)]
+        else:
+            strings_a = _sample_spinless_direct(rdm_a, n_a, shots, rng)
+
+        if n_b == 0:
+            strings_b = [0 for i in range(shots)]
+        elif n_b == norb:
+            strings_b = [sum(1 << j for j in range(norb)) for i in range(shots)]
+        else:
+            strings_b = _sample_spinless_direct(rdm_b, n_b, shots, rng)
+
+        strings_a = restrict_bitstrings(
+            strings_a, orbs_a, bitstring_type=BitstringType.INT
+        )
+        strings_b = restrict_bitstrings(
+            strings_b, orbs_b, bitstring_type=BitstringType.INT
+        )
+
+        if concatenate:
+            strings = concatenate_bitstrings(
+                strings_a,
+                strings_b,
+                BitstringType.INT,
+                length=len(orbs_a),
+            )
+            return convert_bitstring_type(
+                strings,
+                BitstringType.INT,
+                bitstring_type,
+                length=len(orbs_a) + len(orbs_b),
+            )
+
+        return convert_bitstring_type(
+            strings_a,
+            BitstringType.INT,
+            bitstring_type,
+            length=len(orbs_a),
+        ), convert_bitstring_type(
+            strings_b,
+            BitstringType.INT,
+            bitstring_type,
+            length=len(orbs_b),
+        )
+
+
+def _generate_conditionals_unormalized(
+    rdm: np.ndarray, sample: list[int], empty_orbitals: set[int], marginal: float
```

Comment:
You have `empty_orbitals: set[int]` here but it looks like you actually pass it as a list below.